collectors/debugfs.plugin/debugfs_zswap.c

```diff
@@ -0,0 +1,226 @@
+    NETDATA_ZSWAP_REJECTED_COMPRESS_POOR,
+    NETDATA_ZSWAP_REJECTED_KMEM_FAIL,
+    NETDATA_ZSWAP_REJECTED_RALLOC_FAIL,
+    NETDATA_ZSWAP_REJECTED_RRECLAIM_FAIL
+};
+
+
+static struct netdata_zswap_metric zswap_rejected_metrics[] = {
+    {.filename = "/sys/kernel/debug/zswap/reject_compress_poor",
+     .chart_id = "reject_compress_poor",
+     .dimension = "compress_poor",
+     .units = NULL,
+     .title = NULL,
+     .enabled = CONFIG_BOOLEAN_YES,
+     .chart_created = CONFIG_BOOLEAN_NO,
+     .prio = NETDATA_CHART_PRIO_SYSTEM_ZSWAP_REJECTS,
+     .value = -1},
+    {.filename = "/sys/kernel/debug/zswap/reject_kmemcache_fail",
+     .chart_id = "reject_kmemcache_fail",
+     .dimension = "kmemcache_fail",
+     .units = NULL,
+     .title = NULL,
+     .enabled = CONFIG_BOOLEAN_YES,
+     .chart_created = CONFIG_BOOLEAN_NO,
+     .prio = NETDATA_CHART_PRIO_SYSTEM_ZSWAP_REJECTS,
+     .value = -1},
+    {.filename = "/sys/kernel/debug/zswap/reject_alloc_fail",
+     .chart_id = "reject_alloc_fail",
+     .dimension = "alloc_fail",
+     .units = NULL,
+     .title = NULL,
+     .enabled = CONFIG_BOOLEAN_YES,
+     .chart_created = CONFIG_BOOLEAN_NO,
+     .prio = NETDATA_CHART_PRIO_SYSTEM_ZSWAP_REJECTS},
+    {.filename = "/sys/kernel/debug/zswap/reject_reclaim_fail",
+     .chart_id = "reject_reclaim_fail",
+     .dimension = "reclaim_fail",
+     .units = NULL,
+     .title = NULL,
+     .enabled = CONFIG_BOOLEAN_YES,
+     .chart_created = CONFIG_BOOLEAN_NO,
+     .prio = NETDATA_CHART_PRIO_SYSTEM_ZSWAP_REJECTS,
+     .value = -1},
+
+    // The terminator
+    {.filename = NULL, .chart_id = NULL, .enabled = CONFIG_BOOLEAN_NO}
+};
+
+int zswap_collect_data(struct netdata_zswap_metric *metric) {
+    int fd;
+    int ret = 0;
+    char buffer[512];
+
+    char filename[FILENAME_MAX + 1];
+    snprintfz(filename,
+              FILENAME_MAX,
+              "%s%s",
+              netdata_configured_host_prefix,
+              metric->filename);
+    // we are not using profile_open/procfile_read, because they will generate error during runtime.
+    fd = open(filename, O_RDONLY, 0444);
+    if (fd < 0 )
+        return -1;
+
+    ssize_t r = read(fd, buffer, 511);
+    if (r < 0) {
+        ret = -1;
+        goto zswap_charts_end;
+    }
+    // We discard breakline
+    buffer[r - 1] = '\0';
+    metric->value = str2ll(buffer, NULL);
+
+zswap_charts_end:
+    close(fd);
+    return ret;
+}
+
+void zswap_independent_chart(struct netdata_zswap_metric *metric, int update_every) {
+    if (unlikely(!metric->chart_created)) {
+        metric->chart_created = CONFIG_BOOLEAN_YES;
+        fprintf(stdout,
+                "CHART system.%s '' '%s' '%s' 'zswap' '' 'line' %d %d '' 'debugfs.plugin' '%s'\n",
+                metric->chart_id, metric->title, metric->units, metric->prio, update_every, metric->filename);
+        fprintf(stdout, "DIMENSION '%s' '%s' absolute 1 1 ''\n", metric->dimension, metric->dimension);
```

Comment:
I see `absolute` is hard coded. It is 100% wrong, you need to check all the metrics whether are they counters or gauges. Dimension algorithm should be a part of `netdata_zswap_metric`.

Reply:
For zwap you are right, but for extrafrags they are gauge.